pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -122,12 +122,32 @@ func (admitter *VMICreateAdmitter) Admit(ar *admissionv1.AdmissionReview) *admis
 		return webhookutils.ToAdmissionResponse(causes)
 	}
 
+	warnings := warnDeprecatedAPIs(&vmi.Spec, admitter.ClusterConfig)
+	warnings = append(warnings, warnDeprecatedSidecars(vmi)...)
 	return &admissionv1.AdmissionResponse{
 		Allowed:  true,
-		Warnings: warnDeprecatedAPIs(&vmi.Spec, admitter.ClusterConfig),
+		Warnings: warnings,
 	}
 }
 
+func warnDeprecatedSidecars(vmi *v1.VirtualMachineInstance) []string {
```

Comment:
I do not think we should examine annotations, it is a step too far IMO.
Warning about it in the runtime that uses it makes sense, but not at this level.

Reply:
In my opinion, it can be useful and we have precedence with warning over deprecation with featureGates [in the code](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go#L131).

The real downside to me is that it isn't 100% reliable because the *args* are not mandatory, nor the version argument. So, once could be using [passt](https://github.com/kubevirt/kubevirt/tree/main/cmd/sidecars/network-passt-binding) or [slirp](https://github.com/kubevirt/kubevirt/tree/main/cmd/sidecars/network-slirp-binding) would trigger no warning.

The runtime on log works reliably because it happens when the hook sidecar is registering itself.

@vasiliy-ul, wdyt?

Reply:
I still think it's okay to raise awareness of the users. This annotation is part of the deprecated API. Why not raising a warning if it is used? Having only a log message is a good thing, but usually nobody looks at the logs if everything works fine.

Reply:
Another reason is that a user may just change/fix the annotation, and then all this is reversable.

I guess when the deprecation warning will become a strong rejection, you will not want to block changes on such settings at the webhook. As mentioned earlier, it can be fixed.

The fact that these values are not mandatory and defaults kick in, makes this check problematic IMO. I mean, even if the validation here passes, what really matters is the version of the sidecar server.

Reply:
Hmm.. do I understand correctly that with this change the creation and/or change is actually rejected?
Compare to just logging, this is a breaking change, right?

Reply:
This should just print a warning
```
toso@tapioca > k apply -f ./examples/vmi-with-sidecar-hook.yaml
Warning: Hook version v1alpha2 is deprecated. Please use newer versions.
```
Creation/change is not rejected and processed as expected. @victortoso, could you confirm if that is the case?

Reply:
> could you confirm if that is the case?

Yes, user gets a warning but creation works normally.

Reply:
So, @EdDev would rather remove the webhook validation. @vasiliy-ul is in favor to keep it.
I'm more inclined to keep it, even if this is not yet a [perfect](https://github.com/kubevirt/kubevirt/pull/10842#discussion_r1452112389) solution, we can make better heuristics in the future of hooks when promoting the API from alpha to beta. So, my point is just that it feels correct to warn users they are using deprecated sidecars APIs.

@alicefr, @0xFelix, do you have any opinion or suggestion?

Reply:
I like the warning, it is the most immediate thing that user notice. But I agree with Ed when this becomes a rejection then we need to check it on server side

Reply:
I agree that a warning during VM creation is something we want to have. But as @EdDev pointed out, in its current form the check is pretty weak. Considering that the `--version` argument to the sidecar binary is not even required. Can we perhaps make the `--version` argument required on the server binary?

Reply:
Okay, so I'll drop the webhook verification and the check for now. I'll try to think in a possible solution for the next iterations of the hook api.

Thanks for the comments!

Reply:
I do not even understand why it is needed.
The server declares the version it uses and can expose it through the `Info` method.
Ref: https://github.com/kubevirt/kubevirt/blob/80e35d316e24b6a1056c5a551a741554528768b4/pkg/hooks/info/api_info.pb.go#L56

I would expect the client to examine the server version first, then to operate with it in that protocol version.

What is the scenario in which the version needs to be explicitly set as a command line variable, through the annotation?

Reply:
> have new container images running with older clusters and vice versa

Having new images working with old clusters is not something we must support IMO. It is up to the users to keep things in-sync, e.g. upgrade the image only when possible.

The other way is backward compatibility and is a must, just because a cluster upgrade is not going to change the sidecar image.
But this scenario will work fine with the flow mentioned earlier.

Reply:
@alicefr No, we don't have a way to query supported/preferred versions (unless I'm missing something)

Reply:
We could expose the server sidecar hook version in KubeVirt CR. At least in this way, users can read the expected version. Since https://github.com/kubevirt/kubevirt/pull/11001, all authenticated users should be able to read that information.

Reply:
> We could expose the server sidecar hook version in KubeVirt CR. At least in this way, users can read the expected version. Since #11001, all authenticated users should be able to read that information.

The sidecar server runs as part of the sidecar container/image. This is not a cluster global thing.
We can even have multiple versions running in the same virt-launcher pod.
So I am not clear how this suggestion fits exactly.

I do not think webhooks should validate versions, I also do not think it is correct to parse unstructured data from annotations for such validations, especially when the value checks is not required.
I am not even convinced it is relevant to tell the server side with which version it should operate, that should be negotiated between the client and server like most protocol do when a handshake is needed.

There is an opportunity here, once all these issues have been surfaced, to improve the "protocol" and provide a more sustainable solution. I would advice to work on the improvements and prioritize that over patching hacks on the problems seen.

Reply:
> > We could expose the server sidecar hook version in KubeVirt CR. At least in this way, users can read the expected version. Since #11001, all authenticated users should be able to read that information.
> 
> The sidecar server runs as part of the sidecar container/image. This is not a cluster global thing. We can even have multiple versions running in the same virt-launcher pod. So I am not clear how this suggestion fits exactly.
> 

True, but we also support a single version of the virt-launcher image which is the server and it could expose which version of the protocol its support. It could be a way to expose this information in a more friendly way. Right now, the sidecar server API is embedded in virt-launcher. Or is there a way to understand the supported API versions?

The alternative is that users will realize that they need to upgrade the sidecar API version when they fail running the virt-launcher pod.

Do we need to guarantee backward compatibility when the API graduate from alpha to beta?

Reply:
Folks, this is awesome discussion, truly. I'll open a new issue with a tracker of proposed changes for sidecar promotion from alpha and beta and we can probably dig deeper there. I'll just give my opinion on what was raised so far:

> Right now, the sidecar server API is embedded in virt-launcher. Or is there a way to understand the supported API versions?

@alicefr the gRPC server side is actually running in the sidecar. There is a predefined path that they should run their socket and virt-launcher will iterate over all available sockets, doing capabilities query (e.g: what version it is running in that socket and what hooks are handled/exposed).

If we remove, let's say, v1alpha1, the removal will be on virt-launcher, which means that sidecar exposing that version will not be recognized and will fail to run.

> We could expose the server sidecar hook version in KubeVirt CR. At least in this way, users can read the expected version. Since https://github.com/kubevirt/kubevirt/pull/11001, all authenticated users should be able to read that information.

I'm not sure what the best way forward is, we need to brainstorm, but we do need a way to query version and more.

> Do we need to guarantee backward compatibility when the API graduate from alpha to beta?

In my opinion, no. From KubeVirt point of view, I'm mostly worried about not breaking the network pluggins and their 'supported' scenarios. I don't know that, but knowing we can make a plan of how many releases we need to keep a given version deprecated. That's hopefully enough for folks using custom sidecars outside KubeVirt repository and a possible incentive to move minor use cases to use sidecar-shim as, I believe, the plan is that people would only need to rebuild their sidecar with a newer sidecar-shim image as that's focusing on gRPC/api versioning.

> There is an opportunity here, once all these issues have been surfaced, to improve the "protocol" and provide a more sustainable solution. I would advice to work on the improvements and prioritize that over patching hacks on the problems seen.

@EdDev, I agree. I'll update this PR Today and open the tracker issue I mentioned, and we can start discussing ideas for the Beta version.

Thank you all so much for the discussion!

Reply:
Thanks @victortoso for the explication! Sounds good to me.

Reply:
As promised, I opened the following issue to track improvements we want to do for Beta:
https://github.com/kubevirt/kubevirt/issues/11215